### PR TITLE
Add RemoveDir

### PR DESCRIPTION
**Pull request by pedramos**
Necessary addition to implement feature request on aerc: https://todo.sr.ht/~sircmpwn/aerc2/306

In the feature request it is mentioned to pass -f to delete and dir with content but I think that should be part of the application and not the library.

To delete the dir I only deleted the 'cur' folder. According to mbsync (isync) that is the best way to do it.

### CODE REVIEW

**Comment by emersion** [comment]
So this just empties the maildir, rather than deleting it? Maybe we should rename it to `RemoveAll`.

File: maildir.go [hunk 1]
```diff
@@ -413,6 +413,11 @@ func (d Dir) Remove(key string) error {
 	return os.Remove(f)
 }
 
+// RemoveDir removes the actual Maildir Dir
+func (d Dir) RemoveDir() error {
+	return os.RemoveAll(string(d) + "cur")
```
Comment (emersion) [inline maildir.go]: We should re-create the `cur` dir.

File: maildir_test.go [hunk 2]
```diff
@@ -391,3 +391,47 @@ func BenchmarkFilename(b *testing.B) {
 		}
 	}
 }
+
+func TestrRemoveDir(t *testing.T) {
```
Comment (emersion) [inline maildir_test.go]: Typo: "TestRemoveDir"

**Comment by pedramos** [comment]
Sorry, I will amend the typo. Silly mistake 

The feature intends to delete the actual Dir, not to empty it.
The function should allow implementation of rmdir in aerc. Re-creating cur is not intended.

I could have delete the whole thing (cur + tmp + new, and its parent dir) but I believe in that case, software which fetches from IMAP to maildir, will loose track of the dir and won't know it was marked to be deleted.
If that kind of software won't recognize the dir was marked as deleted, it won't propagate the change to the server.

At least it was how I understood it all worked when deleting a maildir dir, maybe I'm seeing it wrong.

Maybe it would make sense though, to delete the content of tmp and new as well.

**Comment by emersion** [comment]
>software which fetches from IMAP to maildir, will loose track of the dir and won't know it was marked to be deleted

Do you have a source?

>The feature intends to delete the actual Dir, not to empty it.

Then it shouldn't leave lingering files and dirs.

**Comment by pedramos** [comment]
I took this impression from mbsync man page:

 >       Remove {None|Master|Slave|Both}
>               Propagate mailbox deletions [to  the  Master/Slave].   Otherwise
>               print  an  error message and skip that mailbox pair if a mailbox
>               does not exist but the corresponding sync state does.
>               For MailDir mailboxes it is sufficient to delete the cur/ subdi‐
>               rectory to mark them as deleted. This ensures compatibility with
>               SyncState *.
>               Note that for safety, non-empty mailboxes are never deleted.
>               (Global default: None)

I searched for this to confirm this is some kind of standard before sending the patch but I couldn't find anything.

> Then it shouldn't leave lingering files and dirs.

I understand with your point, and in general I agree with it. Although, for this specific case, regaring files we can do something about by deleting all of them. For the directories, I understand how it is easy to get this kind of tools confused if we delete the dirs as well because we would loose the "deleted" state.

**Comment by pedramos** [comment]
Now it cleans everything, leaving Dir with empty tmp and new.

File: maildir.go [hunk 3]
```diff
@@ -413,6 +413,29 @@ func (d Dir) Remove(key string) error {
 	return os.Remove(f)
 }
 
+// RemoveDir removes the actual Maildir Dir
+func (d Dir) RemoveDir() error {
+	err := os.RemoveAll(filepath.Join(string(d), "tmp"))
+	if err != nil {
+		return err
+	}
+	os.RemoveAll(filepath.Join(string(d), "new"))
```
Comment (emersion) [inline maildir.go:422]: This is missing error handling

**Comment by emersion** [comment]
The function name is pretty confusing.

Still unsure about the feature itself -- please provide references explaining why this is necessary.

**Comment by emersion** [comment]
Closing because I don't think this feature is worth maintaining.

**Pull request closed without merging**